lib/relooper/src/graph/caterpillar.rs

```diff
@@ -1,166 +1,118 @@
-                )
-            }
-        }
-    };
-    let mut outedges: HashMap<CaterpillarLabel, CfgEdge<CaterpillarLabel>> = cfg
-        .out_edges
-        .iter()
-        .filter(|(label, _)| !label.is_dynamic)
-        .map(to_caterpillar_edge)
-        .collect();
-    let jumpdests: Vec<usize> = cfg
-        .out_edges
-        .iter()
-        .filter_map(|(label, _)| {
-            if label.is_jumpdest {
-                Some(label.cfg_label)
-            } else {
-                None
+pub fn unfold_dyn_edges<T: CfgLabel>(cfg: Cfg<EvmLabel<T>>) -> Cfg<CaterpillarLabel<T>> {
+    let mut cat_cfg = cfg.map_label(|&label| CaterpillarLabel::Original(label.cfg_label));
+
+    let dyn_nodes: Vec<_> = cfg.nodes().into_iter().filter(|l| l.is_dynamic).collect();
+    let jumpdests: Vec<_> = cfg.nodes().into_iter().filter(|l| l.is_jumpdest).collect();
+
+    match &jumpdests[..] {
+        [] => cat_cfg,
+        [single] => {
+            let single_edge = CaterpillarLabel::Original(single.cfg_label);
+            for d in dyn_nodes {
+                cat_cfg.add_edge(CaterpillarLabel::Original(d.cfg_label), CfgEdge::Uncond(single_edge));
             }
-        })
-        .collect();
-    let new_nodes: Vec<CaterpillarLabel> = jumpdests
-        .iter()
-        .enumerate()
-        .map(|(index, jumpdest)| CaterpillarLabel::Generated(index, *jumpdest))
-        .collect();
-    for label in cfg.out_edges.keys() {
-        if !label.is_dynamic {
-            continue;
-        }
-        outedges.insert(
-            CaterpillarLabel::Original(label.cfg_label),
-            CfgEdge::Uncond(new_nodes[0]),
-        );
-    }
-    let node_pairs = new_nodes.iter().zip(new_nodes.iter().skip(1));
-    for (node, next_node) in node_pairs {
-        let offset = match node {
-            CaterpillarLabel::Generated(_id, offset) => offset,
-            CaterpillarLabel::Original(_) => panic!("It must be Generated"),
-        };
-        outedges.insert(
-            *node,
-            CfgEdge::Cond(CaterpillarLabel::Original(*offset), *next_node),
-        );
-    }
-    let offset = match new_nodes.last().unwrap() {
-        CaterpillarLabel::Original(_orig) => {
-            panic!("It must be Generated");
-        }
-        CaterpillarLabel::Generated(_id, offst) => offst,
-    };
-    outedges.insert(
-        *new_nodes.last().unwrap(),
-        CfgEdge::Uncond(CaterpillarLabel::Original(*offset)),
-    );
-    let res: Cfg<CaterpillarLabel> = Cfg::<CaterpillarLabel>::from_edges(
-        CaterpillarLabel::Original(cfg.entry.cfg_label),
-        &outedges,
-    )
-    .unwrap();
-    res
-}
+            cat_cfg
+        },
+        [first, second, rest @ ..] => {
+            let last = (CaterpillarLabel::Original(first.cfg_label), CaterpillarLabel::Original(second.cfg_label));
+
+            let (f_c, f_u) = rest.iter().fold(last, |(c, u), jumpdest| {
+                let new_edge = CfgEdge::Cond(c, u);
```

Comment:
I generally don't understand what's going on here but I think adding edge from node to two different jumpdests shouldn't be here...